src/swoop/api/io.py

```diff
@@ -0,0 +1,107 @@
+import io
+import logging
+
+from minio import Minio
+from minio.error import S3Error
+
+logger = logging.getLogger("uvicorn")
+logger.setLevel(logging.INFO)
+
+
+class IOClient:
+    client = None
+    bucket_name = None
+
+    def __init__(
+        self, s3_endpoint: str, access_key: str, secret_key: str, bucket_name=None
+    ):
+        """Initialize IO Client."""
+        self.client = Minio(
+            s3_endpoint.split("//").pop() if "http" in s3_endpoint else s3_endpoint,
+            access_key=access_key,
+            secret_key=secret_key,
+            secure="http://" not in s3_endpoint,
+        )
+        self.bucket_name = bucket_name
+
+    def get_object(self, object_name: str):
+        """Retrieve from object storage."""
+        object_response = None
+        if (
+            object_name
+            and self.bucket_name
+            and self.client.bucket_exists(self.bucket_name)
+        ):
+            try:
+                response = self.client.get_object(self.bucket_name, object_name)
+                object_response = response.json()
+                logger.debug(f"retrieved object content: {object_response}")
+            except (S3Error, UnboundLocalError) as err:
```

Comment:
The `UnboundLocalError` seems concerning. Where did you run into that?

Reply:
If you try to close the response but you provide a non-existent bucket name or object name the response object appears to be undefined, and then it raises an UnboundLocalError on the finally when it's trying to close the connection. I think I can think of another way of handling that.

Reply:
I think it should be like the block below to allow the else catch the situation where the exception is not triggered.
```
        try:
            response = self.client.get_object(self.bucket_name, object_name)
            object_response = response.json()
            logger.debug(f"retrieved object content: {object_response}")
        except (S3Error, UnboundLocalError) as err:
            logger.debug(err)
        else:
            response.close()
            response.release_conn()
```

Reply:
Then I think then the `UnboundLocalError` doesn't need to be caught anymore, and we should remove it?

I'd still suggest moving the two lines not expected to raise an exception out of the `try` and into the `else`.

Reply:
Fixed.